### DISCUSSION TITLE
Lightline integration?

### DISCUSSION THREAD

**Discussion by Nate-Wilkins** [jcdickinson/wpm.nvim#6] [Q&A] [answered]
This is somewhat of a newbie question but I'm unsure about how to integrate this with vimscript?
I'm new to these Lua plugins and well not sure what to do.

Does this plugin or Lua need to be configured to integrate with other plugins like lightline?

Thanks!

**Comment by FrancoRav** [comment]
I don't use lightline or vimscript, but if you can use the output of a vim command as the component in lightline, you can probably use `lua print(require("wpm").wpm())` which outputs the wpm count (you can replace wpm() by sorted_graph(), historic_graph() or any other wpm function listed on the README.

Try to have a look at adding a component with a function maybe (check this example in the README)
```
let g:lightline = {
      \ 'colorscheme': 'wombat',
      \ 'active': {
      \   'left': [ [ 'mode', 'paste' ],
      \             [ 'gitbranch', 'readonly', 'filename', 'modified' ] ]
      \ },
      \ 'component_function': {
      \   'gitbranch': 'FugitiveHead'
      \ },
      \ }
```
maybe you can create a function that calls the required command and use that (maybe with execute inside the function)

See if you can do something with that. Let me know if you have any issues or it doesn't work, and I'll try to have a look at it.
Again, I don't use lightline or vimscript, so someone else might know better

**Comment by Nate-Wilkins** [comment]
Hey thanks for the help, I really appreciate it!

I tried defining a command in lua like so:

```
vim.api.nvim_create_user_command('DisplayWpm',
  function()
    return "abc"

    -- TODO: Use wpm instead.
    -- return require("wpm").wpm()
  end,
  {})
```

And the lightline configuration looks like this:

```
let g:lightline.component_function = {
  \ 'gitinfo':         'lightline#hunks#composer',
  \ 'filetype':        'lightline#devicon#DevIconFiletype',
  \ 'fileformat':      'lightline#devicon#DevIconFileformat',
  \ 'cocstatus':       'coc#status',
  \ 'currentfunction': 'lightline#coc#CocCurrentFunction',
  \ 'wpm':             'DisplayWpm'
  \ }
```

But I can't get it to recognize the command because I think it's looking for a vimscript function. Doing this does work like I expect:

```
function! DisplayWpm()
  return "wpm:"
endfun
```

But I naively don't know how to call lua functions in here :sweat_smile:

**Comment by FrancoRav** [comment]
Try using execute inside the vim function to execute `lua require("wpm").wpm()`, I think that might work

**Reply by Nate-Wilkins** [reply]
Hm, I tried doing that with:

```vimscript
function! DisplayWpm()
  let l:wpm = lua require("wpm").wpm()
  return "wpm:" ~ l:wpm
endfun
```

And various versions of that but it doesn't seem to work as I thought it would?
I get this error:
```
line  1
E121: Undefined variable :lua
```

I was trying to follow this [nvim-lua-guide](https://github.com/nanotee/nvim-lua-guide#using-lua-from-vimscript) but I think I'm missing something crucial.

**Comment by Nate-Wilkins** [comment, marked as answer]
Well I just had to use `luaeval` correctly... It helps when you read. Thanks for the help @FrancoRav!

This is the final integration:

```vimscript
function! GetWpm()
  return luaeval('string.format("wpm:%s", require("wpm").wpm())')
endfun

" ...

let g:lightline.component_function = {
  \ 'wpm': 'GetWpm'
  \ }
  
" ...
```
